central/complianceoperator/v2/report/manager/manager_impl.go

```diff
@@ -508,38 +508,41 @@ func (m *managerImpl) handleReadyScanConfig() {
 
 func (m *managerImpl) generateReportsFromWatcherResults(result *watcher.ScanConfigWatcherResults) {
 	for _, snapshot := range result.ReportSnapshot {
-		isOnDemand := snapshot.GetReportStatus().GetReportRequestType() == storage.ComplianceOperatorReportStatus_ON_DEMAND
-		if err := m.validateScanConfigResults(result); err != nil {
-			if dbErr := utils.UpdateSnapshotOnError(m.automaticReportingCtx, snapshot, err, m.snapshotDataStore); dbErr != nil {
-				log.Errorf("Unable to upsert the snapshot %s: %v", snapshot.GetReportId(), err)
-			}
-			if isOnDemand {
-				concurrency.WithLock(&m.mu, func() {
-					delete(m.runningReportConfigs, result.ScanConfig.GetId())
-				})
-			}
-			continue
-		}
-		snapshot.GetReportStatus().RunState = storage.ComplianceOperatorReportStatus_PREPARING
-		if err := m.snapshotDataStore.UpsertSnapshot(m.automaticReportingCtx, snapshot); err != nil {
-			log.Errorf("Unable to upsert the snapshot %s: %v", snapshot.GetReportId(), err)
-			if isOnDemand {
-				concurrency.WithLock(&m.mu, func() {
-					delete(m.runningReportConfigs, result.ScanConfig.GetId())
-				})
-			}
-			continue
+		if err := m.generateSingleReportFromWatcherResults(result, snapshot); err != nil {
+			log.Errorf("unable to generate report: %v", err)
 		}
-		generateReportReq := &reportRequest{
-			ctx:                m.automaticReportingCtx,
-			scanConfig:         result.ScanConfig,
-			snapshotID:         snapshot.GetReportId(),
-			notificationMethod: snapshot.GetReportStatus().GetReportNotificationMethod(),
+	}
+}
+
+func (m *managerImpl) generateSingleReportFromWatcherResults(result *watcher.ScanConfigWatcherResults, snapshot *storage.ComplianceOperatorReportSnapshotV2) error {
+	isOnDemand := snapshot.GetReportStatus().GetReportRequestType() == storage.ComplianceOperatorReportStatus_ON_DEMAND
+	defer func() {
+		if isOnDemand {
+			concurrency.WithLock(&m.mu, func() {
+				delete(m.runningReportConfigs, result.ScanConfig.GetId())
+			})
 		}
-		if err := m.handleReportScheduled(generateReportReq, isOnDemand); err != nil {
-			log.Errorf("Unable to handle the report: %v", err)
+	}()
```

Comment:
Is the defer also fine for the execution path that ends with 

```go
	if err := m.handleReportScheduled(generateReportReq, isOnDemand); err != nil {
		return errors.Wrap(err, "unable to handle the report")
	}
```
?

Reply:
If that call fails, then the defer is fine for that path. The problem that I realized is that on the success path we actually need to hold the map until the report is generated [here](https://github.com/stackrox/stackrox/blob/0eb45c795e83390f45fbb5d3329a6fcfd535a533/central/complianceoperator/v2/report/manager/manager_impl.go#L552-L561) and because it is called in a goroutine, the defer does not work for the happy path 😅 .

I have changed the code to move the defer outside of the call to `generateSingleReportFromWatcherResults` in case the call fails.

Reply:
That makes sense! It also looks clean, so I am fine with the current impl. 
Resolving this conversation.